File: week1_decomposition1/1_reachability/Reachability.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Queue;
import java.util.LinkedList;

public class Reachability {
    private static int reach(ArrayList<Integer>[] adj, int x, int y) {
        Queue<Integer> queue = new LinkedList<>();
        boolean[] marked = new boolean[adj.length];
        
        queue.add(x);
        marked[x] = true;

        while( !queue.isEmpty() ) {
            int current = queue.remove();
            if(current == y) {
                return 1;
            }
            for(int neighbour : adj[current]) {
                if( !marked[neighbour] ) {
                    queue.add(neighbour);
                    marked[neighbour] = true;
                }
            }
        }

        return 0;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        System.out.println(reach(adj, x, y));
    }
}

File: week1_decomposition1/1_reachability/reachability.py
#Uses python3

import sys
import queue

def reach(adj, x, y):
    visited = [False] * len(adj)
    q = queue.Queue(maxsize = len(adj))
    q.put(x)
    visited[x] = True

    while not q.empty():
        current = q.get()
        if current == y:
            return 1

        for neighbour in adj[current]:
            if not visited[neighbour]:
                visited[neighbour] = True
                q.put(neighbour)

    return 0

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    x, y = data[2 * m:]
    adj = [[] for _ in range(n)]
    x, y = x - 1, y - 1
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(reach(adj, x, y))
